crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1297,6 +1272,128 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
+                in_reply_to.event = old_in_reply_to.event.clone();
+            }
+        }
+
+        // Start with the canonical case: detect a local timeline item that matches
+        // `event_id` or `transaction_id`.
+        if let Some((local_timeline_item_index, local_timeline_item)) =
+            rfind_event_item(items, |event_timeline_item| {
+                if event_timeline_item.is_local_echo() {
+                    event_id == event_timeline_item.event_id()
+                        || (transaction_id.is_some()
+                            && transaction_id == event_timeline_item.transaction_id())
+                } else {
+                    false
+                }
+            })
+        {
+            trace!(
+                ?event_id,
+                ?transaction_id,
+                ?local_timeline_item_index,
+                "Removing local timeline item"
+            );
+
+            transfer_details(new_event_timeline_item, &local_timeline_item);
+
+            // Remove the local timeline item.
+            return Some(items.remove(local_timeline_item_index));
+        };
+
+        if !settings.vectordiffs_as_inputs {
+            if let Some((remote_timeline_item_index, remote_timeline_item)) =
+                rfind_event_item(items, |event_timeline_item| {
+                    if event_timeline_item.is_remote_event() {
+                        event_id == event_timeline_item.event_id()
+                    } else {
+                        false
+                    }
+                })
+            {
+                trace!(
+                    ?event_id,
+                    ?transaction_id,
+                    ?remote_timeline_item_index,
+                    "Removing remote timeline item (it is a duplicate)"
+                );
+
+                if remote_timeline_item.content.is_redacted()
+                    && !new_event_timeline_item.content.is_redacted()
+                {
+                    warn!("Got original form of an event that was previously redacted");
+                    new_event_timeline_item.content =
+                        new_event_timeline_item.content.redact(&metadata.room_version);
+                    new_event_timeline_item.reactions.clear();
+                }
+
+                transfer_details(new_event_timeline_item, &remote_timeline_item);
+
+                // Remove the remote timeline item.
+                return Some(items.remove(remote_timeline_item_index));
+            }
+        }
+
+        None
+    }
+
+    /// Create a new timeline item from an [`EventTimelineItem`].
+    ///
+    /// It is possible that the new timeline item replaces a duplicated timeline
+    /// event (see [`TimelineEventHandler::deduplicate_local_timeline_item`]) in
+    /// case it replaces a local timeline item.
+    //
+    // Note: this method doesn't take `&mut self` to avoid a borrow checker
+    // conflict with `Self::add_item`. It's fine.
+    fn new_timeline_item(
```

Comment:
Could this be a standalone method, instead of an associated method?

Reply:
Sure.